### PR TITLE
Prevent HGVS_transcript from recalculating predicates unless necessary

### PR DESCRIPTION
**Pull request by aparton**

### CODE REVIEW

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm [hunk 1]
```diff
@@ -1594,6 +1595,16 @@ sub hgvs_protein {
   my $vf = $tv->base_variation_feature;
   my $tr          = $tv->transcript;
 
+  my $adaptor_shifting_flag = 1;
+
+  ## Check previous shift_hgvs_variants_3prime flag and act accordingly
+  $adaptor_shifting_flag = $vf->adaptor->db->shift_hgvs_variants_3prime() if (defined($vf->adaptor) && defined($vf->adaptor->db));
+  
+  my $hash_already_defined = defined($self->{shift_hash});
+  ## Perform HGVS shift even if no_shift is on
```
Comment (ima23) [inline modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: Can you expand the comment to include the adaptor shift option, for e.g. `Perform HGVS shit if adaptor behaviour set for shifting even if transcript no_shift is on`

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm [hunk 2]
```diff
@@ -1650,8 +1662,8 @@ sub hgvs_protein {
   my $ref = $tv->get_reference_TranscriptVariationAllele;
 
   ## Incase the user wants shifted HGVS but not shifted consequences, we run the shifting method  
-  my $hash_already_defined = defined($ref->{shift_hash});
-  $ref->_return_3prime(1) unless $hash_already_defined;
+  my $ref_hash_already_defined = defined($ref->{shift_hash});
```
Comment (ima23) [inline modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: Could you please add a few comments why the hgvs_protein method needs updating? The main PR mentiones HGVS_transcript only. Thanks.

File: modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm [hunk 3]
```diff
@@ -1713,7 +1726,7 @@ sub hgvs_protein {
 
 sub hgvs_offset {
   my $self = shift;
-  return $self->{shift_hash}->{_hgvs_offset};
+  return defined($self->{shift_hash}) ? $self->{shift_hash}->{_hgvs_offset} : $self->{_hgvs_offset};
```
Comment (ima23) [inline modules/Bio/EnsEMBL/Variation/TranscriptVariationAllele.pm]: A comment would be useful, why/when `_hgvs_offset` is part of `shift_hash` and when part of `self`. Thanks.

File: modules/Bio/EnsEMBL/Variation/BaseVariationFeatureOverlapAllele.pm [hunk 4]
```diff
@@ -265,7 +265,7 @@ sub get_all_OverlapConsequences {
     OC: for my $oc (@oc_list) {
 
       last if $assigned_tier && $oc->{tier} > $assigned_tier;
-      my $shifting_offset = defined($self->{shift_hash}) ? $self->{shift_hash}->{shift_length} * $tr->strand : 0;
+      my $shifting_offset = $bvfo->{shifted} && defined($self->{shift_hash}) ? $self->{shift_hash}->{shift_length} * $tr->strand : 0;
```
Comment (ima23) [inline modules/Bio/EnsEMBL/Variation/BaseVariationFeatureOverlapAllele.pm]: A comment about where shifted is set would be useful. For ex `shifted is set in TranscriptVariation`.

**Comment by aparton** [comment]
Currently, hgvs_protein and hgvs_transcript both do not act as fully independent methods. The VariationConsequence pipeline calls these methods in a different order to VEP, and is currently providing different results. This PR should correctly adjust both methods to provide annotations but not incorrect update input $tva objects.

**Comment by ima23** [comment]
merged to `release/100` and `master`

**Pull request merged**